nipio_tests/backend_test.py

```diff
@@ -60,7 +61,87 @@ def test_backend_ends_response_to_A_request_if_ip_is_blacklisted(self):
 
         self._run_backend()
 
-        self._assert_expected_responses(["LOG", "Blacklisted: 127.0.0.2"])
+        self._assert_expected_responses(["LOG", "Rejected IP address: 127.0.0.2"])
+
+    def test_backend_ends_response_to_ANY_request_if_ip_is_not_whitelisted(self):
+        self._send_commands(
+            ["Q", "subdomain.10.0.10.1.lcl.io", "IN", "ANY", "1", "127.0.0.1"]
+        )
+
+        self._run_backend()
+
+        self._assert_expected_responses(["LOG", "Rejected IP address: 10.0.10.1"])
+
+    def test_backend_ends_response_to_A_request_if_ip_is_not_whitelisted(self):
+        self._send_commands(
+            ["Q", "subdomain.10.0.10.1.lcl.io", "IN", "A", "1", "127.0.0.1"]
+        )
+
+        self._run_backend()
+
+        self._assert_expected_responses(["LOG", "Rejected IP address: 10.0.10.1"])
+
+    def test_backend_with_empty_whitelist_responds_to_ANY_request_for_valid_ip(self):
+        self._send_commands(
+            ["Q", "subdomain.10.0.10.1.lcl.io", "IN", "ANY", "1", "127.0.0.1"]
+        )
+
+        backend = self._create_backend()
+        backend.whitelisted_ranges = []
+        backend.run()
+
+        self._assert_expected_responses(
+            ["DATA", "subdomain.10.0.10.1.lcl.io", "IN", "A", "200", "22", "10.0.10.1"],
+            [
+                "DATA",
+                "subdomain.10.0.10.1.lcl.io",
+                "IN",
+                "NS",
+                "200",
+                "22",
+                "ns1.lcl.io",
+            ],
+            [
+                "DATA",
+                "subdomain.10.0.10.1.lcl.io",
+                "IN",
+                "NS",
+                "200",
+                "22",
+                "ns2.lcl.io",
+            ],
+        )
+
+    def test_backend_with_empty_whitelist_responds_to_A_request_for_valid_ip(self):
+        self._send_commands(
+            ["Q", "subdomain.10.0.10.1.lcl.io", "IN", "A", "1", "127.0.0.1"]
+        )
+
+        backend = self._create_backend()
+        backend.whitelisted_ranges = []
+        backend.run()
```

Comment:
Can we extract this out to a method called `_run_backend_without_whitelist` and above as well?